### PR TITLE
Show error messages in create/edit GTAP modal

### LINKED ISSUE #14612: Advanced Sandboxing using questions that return more/other columns than the sandboxed table is not possible anymore, but the errors are not helpful
**Describe the bug**
Advanced Sandboxing using questions that return more/other columns than the sandboxed table is not possible anymore (1.38.0), since it caused an cascade of other problems.
This issue is only about providing helpful errors to admin - not to allow extra columns.

**To Reproduce**
This provides no feedback to the admin, why they cannot select a crazy QB question with extra columns - nothing in the log either.
image

Trying to edit a question that is used for sandboxing from
`SELECT ORDERS.* FROM ORDERS WHERE USER_ID={{sandbox}}`
to
`SELECT ORDERS.*, 'x' AS "EXTRACOL" FROM ORDERS WHERE USER_ID={{sandbox}}`
will now not be possible, since the question metadata is checked before saving.
image

And it's possible to change a QB question and save - in Notebook without ever seen any errors:
image

And if something like this is setup, when upgrading to 1.38.0 there will be errors when users just tries to access tables.
Example: `Output of gtap->source does not match schema: {:source-metadata [nil nil nil nil nil nil (not (map? nil))]}`
image

**Information about your Metabase Installation:**
Commit `6312927` (~0.38.0-rc4)

**Additional context**
Continuation of https://github.com/metabase/metabase-enterprise/issues/524

### PR DESCRIPTION
**Pull request by camsaul**
Fixes (part of?) #14612

GTAP API endpoints weren't actually returning an error message -- just info about the error. GTAP modal on frontend was not displaying errors either way. This PR fixes error messages returned by API endpoints and displays them in the GTAP modal

image

Not sure if this should close #14612 entirely... thoughts @flamber ?

### CODE REVIEW

File: enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx [hunk 1]
```diff
@@ -119,10 +123,21 @@ export default class GTAPModal extends React.Component {
...
+      } else {
+        await GTAPApi.create(gtap);
+      }
+    } catch (error) {
+      console.error("Error saving GTAP", error);
+      const message = error
+        ? error.data
+          ? error.data.message || JSON.stringify(error.data)
+          : JSON.stringify(error)
+        : UNKNOWN_ERROR_MESSAGE;
```
Comment (paulrosenzweig) [inline enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx:138]: `FormMessage` should probably export a function that contains this logic.

File: enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx [hunk 2]
```diff
@@ -119,10 +123,21 @@ export default class GTAPModal extends React.Component {
...
+      this.setState({ error: message });
+      throw new Error(message);
```
Comment (paulrosenzweig) [inline enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx:140]: Why rethrow? To avoid closing the modal?

**Review by paulrosenzweig — APPROVED** [review]
FE looks good to me!

**Comment by flamber** [comment]
Not seeing any messages on CI build of `06eb4a6`. I'm not sure if CI build is not working or I'm not doing the right thing to see the errors.

File: enterprise/backend/test/metabase_enterprise/sandbox/models/group_table_access_policy_test.clj [hunk 3]
```diff
@@ -74,7 +74,7 @@
         (testing "adding new columns = not ok"
```
Comment (dpsutton) [inline enterprise/backend/test/metabase_enterprise/sandbox/models/group_table_access_policy_test.clj:74]: I'm guessing the fix to this is the inclusion of `{:card_id msg}` in the error. If that's what the UI is expecting in order to hook the error up to the card displayed to the user should we have a test checking that that is present?

File: enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj [hunk 4]
```diff
@@ -67,19 +67,23 @@
        ;; `row-level-restrictions` middleware). So include `:type` and `:status-code` information in the ExceptionInfo
        ;; data so it can be passed along if applicable.
        (when-not table-col-base-type
-         (throw (ex-info (tru "Sandbox Cards can''t return columns that aren't present in the Table they are sandboxing.")
-                         {:type        qp.error-type/bad-configuration
-                          :status-code 400
-                          :new-column  col
-                          :expected    (mapv :name table-cols)
-                          :actual      (mapv :name result-metadata-columns)})))
+         (let [msg (tru "Sandbox Cards can''t return columns that aren''t present in the Table they are sandboxing.")]
```
Comment (dpsutton) [inline enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj:70]: `aren''t` double apostrophe

**Comment by dpsutton** [comment]
Just the one comment. I'm not sure what the actual fix was here. The error message already had the string "Sandbox Cards can''t return columns that aren''t present in the Table they are sandboxing." in it. But then you added it to the `errors {:card_id msg}` spot.  If the data is the fix we probably should have a test ensuring that data is correctly put there. I'm also a bit confused by the shape. Associated to a key like `:card_id` I would expect an integer not a string.

File: enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx (hunk 2 above)
Comment (camsaul) [reply enterprise/frontend/src/metabase-enterprise/sandboxes/components/GTAPModal.jsx:140]: yes

File: enterprise/backend/test/metabase_enterprise/sandbox/models/group_table_access_policy_test.clj (hunk 3 above)
Comment (camsaul) [reply enterprise/backend/test/metabase_enterprise/sandbox/models/group_table_access_policy_test.clj:74]: Fixed. This was supposed to be `:message`. @flamber , can you try again

File: enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj (hunk 4 above)
Comment (camsaul) [reply enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj:70]: You have to use double apostrophes in `tru`/trs`. Single ones are for things that are supposed to be returned literally without being quoted or something like that. I put some error checking in the `trs`/`tru` macros that is supposed to catch this but it missed that one for some reason. I noticed it and fixed it here

Comment (dpsutton) [reply enterprise/backend/src/metabase_enterprise/sandbox/models/group_table_access_policy.clj:70]: ah. cool. my bad. :)

**Comment by camsaul** [comment]
@dpsutton the convention we've been using to return error messages from the API is *usually* like 

```clj
{:errors {field-with-error message}}
```

But not everything follows that pattern. In this case I thought `card_id` was the Field most clearly associated with the error. I tried adding that first to see if it would make the error message show up. That ended up not working, so I had to go add FE code to show the error as well.

In a few other places the API just returns a string error message e.g. `"Not found"` or returns a `:message` key. Since I was adding the FE code I ultimately went with using `message` instead of `errors`. I fixed the PR to explicitly include that change. I originally made some tweaks to the API middleware to always include the exception message as `:message`, so it *was* working for me when I took that screenshot, but I reverted that middleware change thinking it wasn't needed. So that's probably why it didn't work for @flamber just now.

It should be fixed now however

**Comment by camsaul** [comment]
@nemanjaglumac, what are the chances you could do me a humongous favor and write a test case to make sure Sandbox error messages show up now in this PR? I think you'd be able to do it a lot quicker than I would

**Comment by nemanjaglumac** [comment]
> @nemanjaglumac, what are the chances you could do me a humongous favor and write a test case to make sure Sandbox error messages show up now in this PR? I think you'd be able to do it a lot quicker than I would

Sure thing.
I'll try to reuse parts of old https://github.com/metabase/metabase-enterprise/issues/524.

**Comment by dpsutton** [comment]
@camsaul that makes sense. thanks for the explanation

**Comment by camsaul** [comment]
The failing Oracle/Vertica Cypress test should be fixed by #14655

**Review by flamber — COMMENTED** [review]
Awesome - gives good error messages to the admin :+1:
Tested commit `a292494`

But if I upgrade from 1.37.8 with a sandbox-card with an extra column, then the users might see this not very helpful error:
`Output of gtap->source does not match schema: [0;33m {:source-metadata [nil nil nil nil nil nil nil nil (not (map? nil))]} [0m `
image

---

The admin can still create/change questions in a way where they don't see any nice error messages, so they can still add columns unintentionally.
Basically the same problem as the reopened part of EE520.

Can change an already connected SQL (adding `, 'x' AS "ExtraCol"`), since the query doesn't run because the variable is non-optional (not specifically with Required set):
```
SELECT *, 'x' AS "ExtraCol" FROM "PRODUCTS" WHERE "CATEGORY"={{sandbox}}
```
The user then has the "ExtraCol" in their view too:
image

It's also possible to create a Notebook question and do stuff (add Custom Column etc) and then save the question without clicking Visualize, but that's an edge case (think it's related to issue 13468).

**Pull request merged**